### PR TITLE
Community wrapper snippets

### PR DESCRIPTION
**Pull request by AndreasArvidsson**
This is the Cursorless side of the community wrapper snippets

https://github.com/talonhub/community/pull/1315

## Checklist

- [-] I have added [tests](https://www.cursorless.org/docs/contributing/test-case-recorder/)
- [-] I have updated the [docs](https://github.com/cursorless-dev/cursorless/tree/main/docs) and [cheatsheet](https://github.com/cursorless-dev/cursorless/tree/main/cursorless-talon/src/cheatsheet)
- [-] I have not broken the cheatsheet

### CODE REVIEW

**Review by pokey — COMMENTED** [review]
This looks fine but what happens if they're on old version of community? Need to make sure we degrade gracefully

Also, I believe there's now a spoken form ambiguity with "funk wrap", no?

I wonder if it would be better for talon to just tell Cursorless where the snippets are using the new state.json file and let it parse them itself so it can handle disambiguating 🤔

**Comment by AndreasArvidsson** [comment]
I think it reasonable to expect people to update community before cursorless. I know of no way of using community lists or captures in cursorless that would not show an error in the log if missing.

The ambiguity would be there if we added the command to community as well. You just cant define the same snippet phrase in both places.

I prefer the rpc solution.

**Comment by pokey** [comment]
> I think it reasonable to expect people to update community before cursorless. I know of no way of using community lists or captures in cursorless that would not show an error in the log if missing.

I completely disagree. We have gone out of our way to make Cursorless talon files easy to update by supporting extensive customization. It is not uncommon for people to go months without updating community, because the fact that the only way to tweak community oftentimes is to directly change it means that upgrades can be quite painful

We should discuss a way to make Cursorless backwards compatible with community. We've done it before, but I believe that was for a new action, so we could easily wrap it in a try-catch

> The ambiguity would be there if we added the command to community as well. You just cant define the same snippet phrase in both places.

Not saying it wouldn't, just pointing it out. Could be surprising if you change your function snippet in community and it doesn't update because Cursorless is preferring the Cursorless one

**Comment by AndreasArvidsson** [comment]
Doesn't that mean that we can never rely on new list or captures from community? I know of no mechanism to solve that.

Yes you cant have the same snippet in both. Personally I have totally disabled cursorless snippets. Getting around this problem might be interesting.

**Comment by pokey** [comment]
> Doesn't that mean that we can never rely on new list or captures from community? I know of no mechanism to solve that.

We could add a context key that we set from community and then match on that?

> Yes you cant have the same snippet in both. Personally I have totally disabled cursorless snippets. Getting around this problem might be interesting.

Yeah would be worth discussing. Esp since we ship with a conflicting snippet 🤔

**Comment by AndreasArvidsson** [comment]
> We could add a context key that we set from community and then match on that?
> 
Unfortunately results in the same problem. If that tag is missing from community because of an older version then cursorless would error in the log.


> Yeah would be worth discussing. Esp since we ship with a conflicting snippet 🤔
Agreed.

**Comment by pokey** [comment]
No not a tag a context key. Do those have to be declared?

**Comment by AndreasArvidsson** [comment]
> No not a tag a context key. Do those have to be declared?

Not sure I follow. Do you mean like `@mod.scope`?

Probably same problem

**Comment by pokey** [comment]
Not sure. You don't explicitly declare the key for a scope in the same way as you declare a tag, so it seems possible that it won't throw an exception if undefined.

If that doesn't work, we could possibly make an onready in cursorless talon that tries to call some action (wrapped in a try-catch) to see whether it exists, and if it does, sets a tag

Anyways we can discuss if none of these options make sense to you; there's usually a way to make things work, with enough elbow grease 🙂

**Comment by pokey** [comment]
update from meet-up:

- asked aegis:

> We introduced a new list in community, and would like to rely on the list in cursorless-talon. However, we expect users to upgrade cursorless-talon more frequently than community, and thus can’t assume that they have this list already. How can we leverage this list when it’s present, without things breaking when it isn’t?

- introduce `user.cursorless_use_community_snippets` tag that will enable wrapping / destinational insertion and disable cursorless snippet wrapping / insertion
- but first, @pokey will review https://github.com/cursorless-dev/cursorless/pull/1879

**Comment by pokey** [comment]
I wonder if it's worth considering sharing the snippets from Talon to Cursorless via `.cursorless/state.json`. Then we could eg show them in the VSCode Cursorless side bar, handle scope-specific snippets, use them for fancy Cursorless stuff like https://github.com/cursorless-dev/cursorless/issues/446 etc

**Comment by AndreasArvidsson** [comment]
@pokey updated

**Review by pokey — COMMENTED** [review]
left a couple comments

File: cursorless-talon/src/snippets.py [hunk 1]
```diff
@@ -166,3 +176,15 @@ def cursorless_wrap_with_snippet(
...
+
+@ctx.action_class("user")
+class UserActions:
+    # The module declaration of this action exists in community
+    def wrap_with_snippet(
+        body: str,
+        target: Any,
+        variable_name: str,
+        scope: str = None,
+    ):
```
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply cursorless-talon/src/snippets.py:189]: Good question. Maybe need to hide this one as well behind a tag?

File: cursorless-talon/src/snippets_community.talon
[the remark this answers is not in the document]
Comment (AndreasArvidsson) [reply cursorless-talon/src/snippets_community.talon:1]: No support for that in community yet

File: cursorless-talon/src/snippets.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (pokey) [reply cursorless-talon/src/snippets.py:189]: Ok let's hide this behind `cursorless_use_community_snippets` tag

**Comment by pokey** [comment]
per https://github.com/talonhub/community/pull/1305#issuecomment-1814286932, we will have cursorless own the actions / rules, and just call the actions from that comment to get what it needs about snippet

**Comment by AndreasArvidsson** [comment]
I still feel that actions gives us more flexibility to call from custom commands.

**Comment by AndreasArvidsson** [comment]
Thanks. Community updated with your suggestion

**Comment by AndreasArvidsson** [comment]
Probably doable?

**Comment by pokey** [comment]
update from meet-up:

- we do want tests
- @AndreasArvidsson will keep this for now, but if time doesn't allow, @pokey will steal this PR in a week

**Comment by pokey** [comment]
Ok I added tests, but:

- See caveat in https://github.com/talonhub/community/pull/1315/files#r1471234714
- It relies on user having `"code"` snippet (code block). That's shipped in community, so should be fine, but we should really mock community snippet dir setting to make this more robust. Prob fine for now tho I think as long as we add a caveat somewhere

**Comment by pokey** [comment]
update from meet-up:

- run one last local test and then ship this alongside https://github.com/talonhub/community/pull/1315

**Review by pokey — APPROVED** [review]

**Comment by brief** [comment]
Is this not merging because of the failing test?

**Comment by pokey** [comment]
Yeah it got caught in that weird windows CI issue; totally unrelated to this PR. Just updated the branch so it gets the fix.   Should pass and merge shortly

**Pull request merged**